crates/rustc_codegen_spirv/build.rs

```diff
@@ -63,21 +73,189 @@ fn check_toolchain_version() -> Result<(), Box<dyn Error>> {
+/// with "legacy" Rust-GPU pointer handling (mainly typed `alloca`s).
+//
+// FIXME(eddyb) get rid of this as soon as it's not needed anymore.
+fn generate_pqp_cg_ssa() -> Result<(), Box<dyn Error>> {
+    let sysroot = rustc_output("--print=sysroot")?;
+    let sysroot = Path::new(sysroot.trim());
+    let cg_ssa_dir = sysroot.join("lib/rustlib/rustc-src/rust/compiler/rustc_codegen_ssa");
+    if !cg_ssa_dir.is_dir() {
+        return Err(format!(
+            "missing `rustc-dev` component from toolchain `{}` (at {})",
+            env::var("RUSTUP_TOOLCHAIN").unwrap_or_else(|_| "<unknown>".into()),
+            sysroot.display(),
+        )
+        .into());
+    }
+
+    let mut cg_ssa_lib_rc_attrs = String::new();
+    let mut cg_ssa_lib_rs_extern_crates = String::new();
+    let writeln = |s: &mut String, line: &str| {
+        *s += line;
+        s.push('\n');
+    };
+
+    let out_dir = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    let out_pqp_cg_ssa_dir = out_dir.join("pqp_cg_ssa");
+    let _ = fs::remove_dir_all(&out_pqp_cg_ssa_dir);
+
+    let mut queue = VecDeque::new();
+    queue.push_back(PathBuf::new());
+    while let Some(suffix) = queue.pop_front() {
+        let in_dir = cg_ssa_dir.join(&suffix);
+        let out_dir = out_pqp_cg_ssa_dir.join(&suffix);
+        fs::create_dir_all(&out_dir)?;
+
+        for entry in in_dir.read_dir()? {
+            let entry = entry?;
+            let relative_path = suffix.join(entry.file_name());
+            if entry.file_type()?.is_dir() {
+                queue.push_back(relative_path);
+                continue;
+            }
+
+            let in_path = entry.path();
+            let out_path = out_dir.join(entry.file_name());
+
+            let mut src = fs::read_to_string(in_path)?;
+
+            // FIXME(eddyb) `regex` crate might be more efficient here.
+            src = src.replace("crate::", "crate::maybe_pqp_cg_ssa::");
+            // HACK(eddyb) needed for `fluent` diagnostic identifiers.
+            src = src.replace("codegen_ssa_", "codegen_spirv_");
+
+            if relative_path == Path::new("src/lib.rs") {
+                // HACK(eddyb) rewrite line-by-line to comment (and copy) some lines.
+                for line in mem::take(&mut src).lines() {
+                    if line.starts_with("#!") {
+                        src += "// ";
+                        if !line.starts_with("#![doc(") && line != "#![warn(unreachable_pub)]" {
+                            writeln(&mut cg_ssa_lib_rc_attrs, line);
+                        }
+                    } else if line == "#[macro_use]" || line.starts_with("extern crate ") {
+                        src += "// ";
+                        writeln(&mut cg_ssa_lib_rs_extern_crates, line);
+                    }
+                    writeln(&mut src, line);
+                }
+            }
+
+            // HACK(eddyb) "typed alloca" patches.
+            if relative_path == Path::new("src/traits/builder.rs") {
```

Comment:
Should these just be .patch files that are then applied rather than being in the source?

Reply:
`str::replace` may be expensive but I don't want to deal with adding a whole extra dependency just to rewrite one line in each of 3 files (if there is even a good one - keep in mind this has to be inherently cross-platform).

I didn't even bother with the `regex` crate because this worked. Maybe if it's a bottleneck we could try to fix that but I don't want to make this stuff too cushy (ideally we can nuke it just after we release 0.10).